DistributedRectilinearMeshGenerator for PolycrystalVoronoi

fdkong opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

Closes #16143

fdkong committed:
Multible side layers

Added a relationship manager into DistributedRectilinearMeshGenerator

This help reserve user-specified layers of ghosting elements

fdkong committed:
Added a test when using PolycrystalUserObjectBase UO

Closes #16143

moosebuild commented:
Job [Private App tests PETSc alt](https://civet.inl.gov/job/619429/) on efbb30e : invalidated by @loganharbour

Test box failure

moosebuild commented:
Job [Test debug](https://civet.inl.gov/job/619418/) on efbb30e : invalidated by @loganharbour

Test box failure

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/619424/) on efbb30e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16144/site)

This comment will be updated on new commits.

fdkong committed:
Move tests into generator in subdir distributed_rectilinear_mesh_generator

fdkong committed:
Moved dmg_displaced_mesh into distributed_rectilinear_mesh_generator

fdkong committed:
Renamed distributed_rectilinear_mesh_generator to distributed_rectilinear

lindsayad commented:
@fdkong are you sure that a relationship manager is the answer here? You should be hitting this code:
```c++
void
MooseApp::attachRelationshipManagers(MeshBase & mesh, MooseMesh & moose_mesh)
{
  for (auto & rm : _relationship_managers)
  {
    if (rm->isType(Moose::RelationshipManagerType::GEOMETRIC))
    {
      if (rm->attachGeometricEarly())
      {
        rm->init(mesh);
        mesh.add_ghosting_functor(*rm);
      }
      else
      {
        // If we have a geometric ghosting functor that can't be attached early, then we have to
        // prevent the mesh from deleting remote elements
        moose_mesh.allowRemoteElementRemoval(false);

        if (const MeshBase * const moose_mesh_base = moose_mesh.getMeshPtr())
        {
          if (moose_mesh_base != &mesh)
            mooseError("The MooseMesh MeshBase and the MeshBase we're trying to attach "
                       "relationship managers to are different");
        }
        else
          // The MeshBase isn't attached to the MooseMesh yet, so have to tell it not to remove
          // remote elements independently
          mesh.allow_remote_element_removal(false);
      }
    }
  }
}
```
So you *should not be* deleting any remote elements when `prepare_for_use` is called. You should not be deleting any remote elements until:
```c++
  else if (_current_task == "delete_remote_elements_after_late_geometric_ghosting")
  {
    if (_displaced_mesh &&
        (_mesh->needsRemoteElemDeletion() != _displaced_mesh->needsRemoteElemDeletion()))
      mooseError("Our reference and displaced meshes are not in sync with respect to whether we "
                 "should delete remote elements.");

    // We currently only trigger the needsRemoteDeletion flag if somebody has requested a late
    // geometric ghosting functor and/or we have a displaced mesh
    if (_mesh->needsRemoteElemDeletion())
    {
      _mesh->deleteRemoteElements();
      if (_displaced_mesh)
        _displaced_mesh->deleteRemoteElements();
    }
  }
```
The `dlete_remote_elements_after_late_geometric_ghosting_task` doesn't happen until well after addition of periodic boundaries and addition of "late" geometric RMs.

I'm not going to rule out the existence of a bug, but that's how the code *should* be operating now. Are you sure that your problem isn't that you simply *never have* the elements needed for the grain tracker when doing your distributed mesh generation? Ghosting functors just prevent deletion of existing elements; they never restore any elements.

fdkong committed:
Skip geometric proxy relationship manager

fdkong commented:
> @fdkong are you sure that a relationship manager is the answer here?

I can not say it is a relationship manager bug but the issue is actually related to RM.

1. If geometric RM is chosen as not attach early, then that geometric RM will not be attached any more for the entire simulation. For example, geometric proxy RMs never be attached in the current code base. It does not cause problems because there is an equivalent  algebraic RMs in there always.

2. If a geometric RM is attached early, and then it will be initialized without knowing there are some Periodic BCs because it was initialized too early.  Geo RM with PBCs and without PBCs should give different geometric information.

> You should be hitting this code: `MeshGenerator::buildDistributedMesh` -> `MooseMesh::buildTypedMesh<DistributedMesh> ` ->
> 
> ```c++
> void
> MooseApp::attachRelationshipManagers(MeshBase & mesh, MooseMesh & moose_mesh)
> {
>   for (auto & rm : _relationship_managers)
>   {
>     if (rm->isType(Moose::RelationshipManagerType::GEOMETRIC))
>     {
>       if (rm->attachGeometricEarly())
>       {
>         rm->init(mesh);
>         mesh.add_ghosting_functor(*rm);
>       }
>       else
>       {
>         // If we have a geometric ghosting functor that can't be attached early, then we have to
>         // prevent the mesh from deleting remote elements
>         moose_mesh.allowRemoteElementRemoval(false);
> 
>         if (const MeshBase * const moose_mesh_base = moose_mesh.getMeshPtr())
>         {
>           if (moose_mesh_base != &mesh)
>             mooseError("The MooseMesh MeshBase and the MeshBase we're trying to attach "
>                        "relationship managers to are different");
>         }
>         else
>           // The MeshBase isn't attached to the MooseMesh yet, so have to tell it not to remove
>           // remote elements independently
>           mesh.allow_remote_element_removal(false);
>       }
>     }
>   }
> }
> ```
> 
> So you _should not be_ deleting any remote elements when `prepare_for_use` is called. You should not be deleting any remote elements until:
> 
> ```c++
>   else if (_current_task == "delete_remote_elements_after_late_geometric_ghosting")
>   {
>     if (_displaced_mesh &&
>         (_mesh->needsRemoteElemDeletion() != _displaced_mesh->needsRemoteElemDeletion()))
>       mooseError("Our reference and displaced meshes are not in sync with respect to whether we "
>                  "should delete remote elements.");
> 
>     // We currently only trigger the needsRemoteDeletion flag if somebody has requested a late
>     // geometric ghosting functor and/or we have a displaced mesh
>     if (_mesh->needsRemoteElemDeletion())
>     {
>       _mesh->deleteRemoteElements();
>       if (_displaced_mesh)
>         _displaced_mesh->deleteRemoteElements();
>     }
>   }
> ```
> 
> The `dlete_remote_elements_after_late_geometric_ghosting_task` doesn't happen until well after addition of periodic boundaries and addition of "late" geometric RMs.

Yes, deletion happened at the right time. The issue is that the geometric RM does not know there are PBCs since the geometric RM is initialized too early. 
> 
> I'm not going to rule out the existence of a bug, but that's how the code _should_ be operating now. Are you sure that your problem isn't that you simply _never have_ the elements needed for the grain tracker when doing your distributed mesh generation?

I checked this by manually adding `return` into the beginning of `DistributedMesh::delete_remote_elements` to stop libmesh from deleting anything. And then the whole world is happy. Definitely, this is not a fix. That is used for checking if I generate enough ghosting elements at the first place.

> Ghosting functors just prevent deletion of existing elements; they never restore any elements.

lindsayad commented:
> I checked this by manually adding `return` into the beginning of `DistributedMesh::delete_remote_elements` to stop libmesh from deleting anything. And then the whole world is happy. Definitely, this is not a fix. That is used for checking if I generate enough ghosting elements at the first place.

Ah ok very good to know :+1: 

> Yes, deletion happened at the right time. The issue is that the geometric RM does not know there are PBCs since the geometric RM is initialized too early.

Have you seen this code at the end of `AddPeriodicBCAction.C`:
```c++
    // Now make sure that the mesh default ghosting functor has its periodic bcs set
    _mesh->getMesh().default_ghosting().set_periodic_boundaries(
        nl.dofMap().get_periodic_boundaries());
    if (displaced_problem)
      displaced_problem->mesh().getMesh().default_ghosting().set_periodic_boundaries(
          displaced_problem->nlSys().dofMap().get_periodic_boundaries());
```
I had to do this to get some of your more recent dmg tests with periodic boundaries to work. I am curious why it is not working for your current case?

fdkong committed:
Added tests for multiple layers of ghosting elements

in distributed generator

Close #16143

fdkong committed:
Moved ElemSideNeighborLayers*.{C,h} from test to framework

fdkong committed:
Skip remote neighbors of evaluable elements in PolycrystalUserObjectBase.C and FeatureFloodCount.C

In general, {evaluable elements} >= {local elements} U {algebraic ghosting elements}. That is,
the number of evaluable elements does NOT necessarily equal to the number of local and algebraic
ghosting elements. The neighbors of evaluable elements can be remote even though we have two layers
of geometric ghosting elements.

Closes #16143

fdkong committed:
Added test for outputting ghosting and evaluable elements when using distributed generator

Closes #16143

fdkong committed:
Added missing gold files

fdkong committed:
Regold test/tests/relationship_managers/two_rm because the original

solution is just wrong

fdkong commented:
Regold `relationship_managers/two_rm.evaluable_neighbors/replicated ` because the original gold is just so wrong

In this PR (two layers of geometric ghosts):

![geometric_ghost](https://user-images.githubusercontent.com/10508374/102532576-68d92a00-4061-11eb-8b43-814a038129bb.jpeg)


Before this PR (two layers of element ghosts):

![geometric_ghost_old](https://user-images.githubusercontent.com/10508374/102532614-742c5580-4061-11eb-9a73-1de56823d482.jpeg)

fdkong committed:
Regold dgkernels/2d_diffusion_dg.no_additional_rms

We attach "late" RM now.

New:

Relationship Managers:
  Geometric and algebraic and coupling:ElementSideNeighborLayers (1 layer) for regular_dg_diffusion_nl, coupling_functor_check_nl
  Algebraic:               ElementSideNeighborLayers (1 layer) for regular_dg_diffusion_aux
  Geometric:               GhostPointNeighbors for exodus
  Default:                 libMesh::DefaultCoupling x 2
  Default:                 libMesh::GhostPointNeighbors

Old:

Relationship Managers:
  Geometric and algebraic and coupling:ElementSideNeighborLayers (1 layer) for regular_dg_diffusion_nl, coupling_functor_check_nl
  Algebraic:               ElementSideNeighborLayers (1 layer) for regular_dg_diffusion_aux
  Default:                 libMesh::DefaultCoupling x 2
  Default:                 libMesh::GhostPointNeighbors

fdkong committed:
Regold dgkernels/2d_diffusion_dg.proper_ghosting_with_action_parallel

We attach "late" geometric RM correctly

New:

Relationship Managers:
  Geometric:               GhostPointNeighbors for exodus
  Algebraic:               ElementSideNeighborLayers (1 layer) for coupling_functor_check_nl
  Algebraic:               ElementSideNeighborLayers (1 layer) for coupling_functor_check_aux
  Coupling:                ElementSideNeighborLayers (1 layer) for coupling_functor_check_nl
  Default:                 libMesh::DefaultCoupling x 2
  Default:                 libMesh::GhostPointNeighbors

Old:

Relationship Managers:
  Algebraic:               ElementSideNeighborLayers (1 layer) for coupling_functor_check_nl
  Algebraic:               ElementSideNeighborLayers (1 layer) for coupling_functor_check_aux
  Coupling:                ElementSideNeighborLayers (1 layer) for coupling_functor_check_nl
  Default:                 libMesh::DefaultCoupling x 2
  Default:                 libMesh::GhostPointNeighbors

fdkong committed:
Regold dgkernels/2d_diffusion_dg.proper_ghosting_with_action_serial

We attach "late" geometric RM correctly

New:

Relationship Managers:
  Geometric:               GhostPointNeighbors for exodus
  Coupling:                ElementSideNeighborLayers (1 layer) for coupling_functor_check_nl
  Default:                 libMesh::DefaultCoupling x 2
  Default:                 libMesh::GhostPointNeighbors

Old:

Relationship Managers:
  Coupling:                ElementSideNeighborLayers (1 layer) for coupling_functor_check_nl
  Default:                 libMesh::DefaultCoupling x 2
  Default:                 libMesh::GhostPointNeighbors

lindsayad commented:
I see...because before we were not even attaching geometric ghosting functors if the mesh was replicated

moosebuild commented:
Job [Modules debug](https://civet.inl.gov/job/649775/) on d26929c : invalidated by @fdkong

fdkong committed:
Added missing md files

lindsayad commented on framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C:
```diff
@@ -54,6 +54,26 @@ DistributedRectilinearMeshGenerator::validParams()
+      2,
+      "Number of layers of off-processor side neighbors is reserved during mesh generation");
+
+  params.addRelationshipManager(
+      "ElementSideNeighborLayers",
+      Moose::RelationshipManagerType::GEOMETRIC,
+      [](const InputParameters & obj_params, InputParameters & rm_params) {
+        // Let this RM safeguard users specified ghosted layers
+        rm_params.set<unsigned short>("layers") = obj_params.get<unsigned>("num_side_layers");
+        // We can not attach geometric early here because some simulation related info, such as,
+        // periodic BCs is not available yet during an early stage. Periodic BCs will be passed
+        // into ghosting functor during initFunctor of ElementSideNeighborLayers. There is no hurt
+        // to attach geometric late for general simulations that do not have extra requirements on
+        // ghosting elements. That is especially true distributed generated meshes since there is
+        // not much redundant info.
+        rm_params.set<bool>("attach_geometric_early") = false;
+      });
```
I'm curious whether this is necessary. When I think about a ghosting need, I (almost) always think there's an algebraic or coupling driver. For instance I wouldn't request that my mesh keep 10 layers of neighbors ghosted just for fun; I would only do so if I plan on accessing vector or matrix entries that are 10 layers out. It seems that the motivator for this PR is PolycrystalVoronoi. If I look in `PolycrystalUserObjectBase`, I see that we should already be adding two layers of neighbor side geometric ghosting. I'm guessing that you still probably need your `num_side_layers` parameter here, but I'm just thinking you can probably leave off addition of the geometric RM here. I might even suggest making `num_side_layers` a required parameter such that users have to think about what they are doing, e.g. what objects do they have in their simulation that require off-process side neighbors.

fdkong replied:
`PolycrystalVoronoi` or `grain tracker` requires two layers of geometric ghosting elements (by default), but we need only one layer of coupling and algebraic elements.   That being said, there are use cases where a geometric RM can not be overridden by other coupling and algebraic RMs. 

I do like keeping `the addition of the geometric RM` in the mesh generator.   If there is no addition of a geometric RM here, we can not implement `num_side_layers` correctly. For example, if I have `num_side_layers=3`, then go to run the diffusion.i, and then we will end up with results of `num_side_layers=1`.

lindsayad replied:
Ok, I am fine with keeping the geometric RM addition here. I'm curious: is there a problem if a user specifies `num_side_layers=1` in the mesh generator but then they run `PolyCrystalVoronoi` or the grain tracker? If I understand correctly `num_sides_layers` controls how many side layers of elements you *generate* so if you don't generate sufficient side layers, then you will be in trouble?

If my above scenario is indeed a problem, then I don't have any bright idea for fixing it automatically for the user. At that point maybe we just say it is the user's fault.

fdkong replied:
Yes, good point. `num_sides_layers` is used to control we have sufficient ghosting elements at the first place. 

That is an egg-chicken loop here. Without a mesh, we can not create kernels, UOs, etc. Without knowing RMs in kernels, MOs, etc. we have no idea how much information we need to generate in the first place.

Looking forward someday we will have an intelligent  RM that can grow itself automatically.  If such an intelligent  RM exists, then we do not even to generate overlapping domain at the first place

lindsayad reviewed (changes requested):
Glad we are continuing to improve our relationship manager capability!

fdkong committed:
Update framework/doc/content/source/userobject/ElemSideNeighborLayersGeomTester.md

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

fdkong commented on framework/doc/content/source/auxkernels/ElementUOAux.md:
```diff
@@ -0,0 +1,19 @@
+ ghosting information.
+
+
+!! Describe and include an example of how to use the ElementUOAux object.
```
Haha, I will remove this

fdkong committed:
Update framework/src/base/MooseApp.C

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

fdkong commented on framework/src/base/MooseApp.C:
```diff
@@ -2020,18 +2026,32 @@ MooseApp::attachRelationshipManagers(MeshBase & mesh, MooseMesh & moose_mesh)
         MooseMesh * const mesh = _action_warehouse.mesh().get();
 
-        if (!rm->attachGeometricEarly())
+        // "attach_geometric_rm_final = true" inidicate that it is the last chance to attach
+        // geometric RMs. Therefore, we need to attach them.
+        if (!rm->attachGeometricEarly() && !attach_geometric_rm_final)
```
Yes, you are right. If there is a pure-geometric RM with `attach_geometri_early=false`, and then this RM never get attached using the old logic.

fdkong commented on framework/src/base/MooseApp.C:
```diff
@@ -2020,18 +2026,32 @@ MooseApp::attachRelationshipManagers(MeshBase & mesh, MooseMesh & moose_mesh)
     {
       if (rm_type == Moose::RelationshipManagerType::GEOMETRIC)
       {
+        // We do not need to attach geometric ProxyRelationshipManager since
+        // all regular relationship managers are added to both undisplaced meshes
+        // and displaced mesh. BTW, geometric ProxyRelationshipManager was not
+        // attached anywhere in the old code base.
+        if (dynamic_cast<ProxyRelationshipManager *>(rm.get()))
+          continue;
```
I think It is a good idea NOT, at the first place, to create geometric proxy relationship managers. It is why I like your idea in your another PR where you are trying to implement "`>=`" operators.  If `>=` is implemented correctly, we can significantly reduce the number of RMs and just keep the necessary  ones.

fdkong replied:
The geometric proxy relationship managers were not added at all using the old logic. 

With the new logic (I added attach geometric RMs late ), geometric proxy relationship managers are then added and that have some side effects.

lindsayad replied:
Yea, let's just not ever add them at all, and then we can remove this `dynamic_cast`

fdkong commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -1602,9 +1602,16 @@ FeatureFloodCount::visitElementalNeighbors(const Elem * elem,
+       * algebraic ghosting elements. The neighbors of evaluable elements can be remote even though
+       * we have two layers of geometric ghosting elements.
+       */
+      if (neighbor_ancestor->is_remote())
```
It is the same.

fdkong commented on framework/src/base/MooseApp.C:
```diff
@@ -1951,11 +1952,16 @@ MooseApp::addRelationshipManager(std::shared_ptr<RelationshipManager> relationsh
-    return false;
+  // We prefer to always add geometric RMs. There is no hurt to add RMs for replicated mesh
+  // since MeshBase::delete_remote_elements{} is not implemented in replicated mesh.
+  // The motivation here is that mesh parallel type might not be set yet for distributed mesh
+  // generator, and for the default parallel type "isDistributedMesh()" is false. All geometric
+  // RMs are just ignored for distributed mesh generator
```
The geometric RMs were ignored for distributed mesh generator because of the old logic here:

```
if (!_action_warehouse.mesh()->isDistributedMesh() && ....)
  return false
```

`_action_warehouse.mesh()->isDistributedMesh()`  was false because `MooseApp::addRelationshipManager` was called before we informed moose that the future mesh was a distributed mesh. 

We deleted this old logic in this PR so that we have all geometric  RMs added for the distributed mesh generator

lindsayad replied:
Yes very good. I am glad we add geometric ghosting for both replicated and distributed now. Uniformity is nice

fdkong committed:
Reword the descriptions in ElemSideNeighborLayersTester.md and ElemSideNeighborLayersGeomTester.md

fdkong commented on framework/doc/content/source/userobject/ElemSideNeighborLayersTester.md:
```diff
@@ -0,0 +1,14 @@
+# ElemSideNeighborLayersTester
+
+!syntax description /UserObjects/ElemSideNeighborLayersTester
+
+## Overview
+
+A user object is used to calculate ghost elements. `element_side_neighbor_layers` is employed to specify the number of layers of ghosting elements.
+
+
+!syntax parameters /UserObjects/ElemSideNeighborLayersTester
+
+!syntax inputs /UserObjects/ElemSideNeighborLayersTester
+
+!syntax children /UserObjects/ElemSideNeighborLayersTester
```
There is a minor difference:

`ElemSideNeighborLayersTester`:

```
  params.addRelationshipManager(
      "ElementSideNeighborLayers",
      Moose::RelationshipManagerType::GEOMETRIC | Moose::RelationshipManagerType::ALGEBRAIC,

      [](const InputParameters & obj_params, InputParameters & rm_params) {
        rm_params.set<unsigned short>("layers") =
            obj_params.get<unsigned short>("element_side_neighbor_layers");
      }

  );
```
`ElemSideNeighborLayersGeomTester`:

```
  params.addRelationshipManager(
      "ElementSideNeighborLayers",
      Moose::RelationshipManagerType::GEOMETRIC,

      [](const InputParameters & obj_params, InputParameters & rm_params) {
        rm_params.set<unsigned short>("layers") =
            obj_params.get<unsigned short>("element_side_neighbor_layers");
      }

  );
```

That being said, `ElemSideNeighborLayersTester` prints ghosting elements that are preserved via either a geometric RM or an algebraic RM, while `ElemSideNeighborLayersGeomTester` prints ghosting elements that are pressers via only a geometric RM.

lindsayad replied:
Can we put this line:

> ElemSideNeighborLayersTester prints ghosting elements that are preserved via either a geometric RM or an algebraic RM

in the `ElemSideNeighborLayersTester` markdown file then? Very helpful sentence.

fdkong replied:
Yes

fdkong committed:
Improve docs in FEProblemBase.h and MooseApp.C

fdkong committed:
Do not create a geometric proxy RM in CreateDisplacedProblemAction

fdkong committed:
Changed 'num_side_layers' to RangeCheckedParam (DistributedRectilinearMeshGenerator.C)

fdkong commented:
OK, I need to rebase on next to resolve conflict.

fdkong commented:
OK, https://github.com/idaholab/moose/pull/16397/files broke the whole world

loganharbour commented:
> OK, https://github.com/idaholab/moose/pull/16397/files broke the whole world

I disagree: https://civet.inl.gov/event/61701/

fdkong commented:
> > OK, https://github.com/idaholab/moose/pull/16397/files broke the whole world
> 
> I disagree: https://civet.inl.gov/event/61701/

Haha, I was talking about different things. What I meant was that this PR (#16144) works well without using PR #16397, but it fails when using #16397. That being said, PR (#16144) workes well based on devel, but fails based on next.


Anyway, I will figure it out

loganharbour commented:
Okay, so YOU broke the world :-)

lindsayad commented:
Together we broke the world. Sorry @fdkong !

lindsayad commented:
We could potentially revert the `RelationshipManager` work from #16397 if you cannot easily figure out the issue. The `RelationshipManager` work turned out not to be essential to the central concept of that PR, but I figured I would keep in stuff like the `operator>=` stuff because I thought it was an improvement

fdkong committed:
We can not rely on dofmap for periodic BCs because dof_map is set by only coupling RM

We should allow geometric and algebraic RMs to setup periodic BCs as well

fdkong commented:
I think I figured it out. 

We do not really call `setDofMap()` for geometric and algebraic RMs. Then the ghosting elements associated with periodic BCs  were ignored for geometric and algebraic  RMs. It seems that we call `setDofMap()` only when RM is coupling.

lindsayad commented:
Nice catch. Here is my conceptual issue: it seems not quite right to always use the nonlinear `DofMaps` periodic boundaries for all algebraic ghosting. For example what if theoretically the aux system `DofMap` had different periodic boundaries or even if we had some other `System/DofMap` in `EquationSystems` which had different periodic boundaries. Then we would not be applying the correct algebraic ghosting. I think the chances of this are small: e.g. I don't think we ever do periodic boundaries for auxiliary variables. But this is me thinking out loud. I guess a solution I would like better would be to make sure that we are calling `setDofMap` correctly in `MooseApp::attachRelationshipManagers` for algebraic (non-coupling) RMs. But I know you have other things to do, so if we don't do that in this PR I'll just create an issue.

fdkong commented:
> But this is me thinking out loud. I guess a solution I would like better would be to make sure that we are calling setDofMap correctly in MooseApp::attachRelationshipManagers for algebraic (non-coupling) RMs. But I know you have other things to do, so if we don't do that in this PR I'll just create an issue.

I think we need to call  `setDofMap` for all RMs regardless of type. 

As we talked earlier, we need two layers (by default) of geometric ghosting elements for grain tracker, while one layer of coupling elements for nonlinear systems.

lindsayad commented:
You’re definitely onto something there. In theory a geometric ghosting functor should consider the union of all periodic boundaries from all DofMaps associated with the EquationSystems associated with the mesh.

lindsayad commented:
A general RM perhaps should have a set of DofMaps

fdkong commented:
OK, it is way more complicated than I thought.  So I would like to do that change in another PR.

lindsayad commented:
Thinking more about it, I think that a `RelationshipManager` should have 0 `DofMaps` if it is meant for pure geometric ghosting, and exactly 1 `DofMap` if it does either algebraic or coupling. If we have an RM in the latter case, the case where we are doing algebraic or coupling ghosting, and that RM is meant to do ghosting for periodic boundaries, then that RM absolutely should do geometric ghosting as well. In this way you would setup a union of `RelationshipManagers` that would geometrically ghost any periodic boundaries needed for algebraic/coupling purposes. Periodic boundaries are intrinsically an algebraic concept so I think it is appropriate to do things this way.

If there is agreement on this, then the only changes that would have to be made would be to also `setDofMap` for `RelationshipManagers` that do algebraic ghosting. And then I would suggest that in any `RelationshipManager` where we `set_periodic_boundaries` (like in `ElementSideNeighborLayers`), we assert that the `RelationshipManager` is *also* of type `GEOMETRIC`.

lindsayad commented:
@fdkong can you look at https://github.com/idaholab/moose/pull/16619? If you like it, maybe you can just cherry-pick onto this PR?

fdkong commented:
Sorry, I did not explain this clear. Before we pu too much energy into this, let us back to the original case I was talking about. For PF simulations with `grain tracker` or `PolycrystalVoronoi`, it need two layers of geometric ghosting elements (geometric RM), one layer of algebraic ghosting elements (precisely, evaluable elements) (algebraic RM), and zero layer of coupling ghosting elements (coupling RM). Let us show a toy example

Partition the mesh into three subdomains for parallel computations

![partition_pbc](https://user-images.githubusercontent.com/10508374/103804547-891c5900-500f-11eb-8c80-4ec089cf83bd.jpeg)


Two layers of geometric ghosting elements for subdomain 0. You could see that if a geometric RM does not know dofmap, then it does not know anything about PBCs, and then top red elements are not going to be ghosted.

![ghosting_0_pbc](https://user-images.githubusercontent.com/10508374/103804850-f4662b00-500f-11eb-8ed3-bad0dcae37d0.jpeg)

Evaluable elements (= local elements + one layer of algebraic ghosting elements) for subdomain 0

![evaluable_0_pbc](https://user-images.githubusercontent.com/10508374/103805029-3abb8a00-5010-11eb-9a3e-0e11295e192f.jpeg)


In `grain tracker` or `PolycrystalVoronoi`, we have a simple logic like this:

```
for (auto element: evaluable_elemtns)
  vist(element's neighbors)
```

Right now, PBCs are imbedded in `dofmap`. Without `dofmap`, I do not know how to tell a geometric RM that it has PBCs

lindsayad commented:
Why is it that you need two layers of geometric ghosting but only one layer of algebraic ghosting for this thing? If they had the same number of layers, that would fix the `DofMap` issue...you would just have one RM that does both geometric and algebraic and because of the latter a `DofMap` would fit naturally

lindsayad commented:
Just as a `MeshBase` does not have `EquationSystems` data, it seems weird for a pure geometric RM to have `DofMap` data

lindsayad commented:
Oh I see: you need two geometric layers because of this code...?

```
for (auto element: evaluable_elemtns)
  visit(element's neighbors)
```

lindsayad commented:
Sigh. I think what I would like to see in the long-term is that when you add periodic boundaries to `DofMap`, they get added to the corresponding `MeshBase` at the same time. But that would be a libmesh change that we probably don't want to wait on

fdkong commented:
> Why is it that you need two layers of geometric ghosting but only one layer of algebraic ghosting for this thing?

Good question. If we assume that the number of layers of geometric ghosting is always the same as  that of algebraic  
ghosting, then we do not even need any geometric RM at the first place because the algebraic RM cover everything we need.

 But I think it is legal to ask for more geometric elements than algebraic elements. 

For `grain tracker` or `PolycrystalVoronoi`, as-is code does need two layers of geometric elements 

```
for (auto element: evaluable_elemtns)
   visit(element's neighbors)
```

We might be able to get rid of this  requirement someday by creating an intelligent feature flooding count algorithm , but a good factorization of `grain tracker` takes even more efforts. 

We do not really need `dofmap`,  but we do need to have a geometric RM understand PBCs

fdkong commented:
> Oh I see: you need two geometric layers because of this code...?
```
for (auto element: evaluable_elemtns)
  visit(element's neighbors)
```

Yes, it is right

lindsayad reviewed (approved):
Ok I think PR is accomplishing what it sought out to do. I would like to add more robustness such that we can revert https://github.com/idaholab/moose/pull/16144/commits/d84bf9569e590aa9caac76542670ade9a553eb57, but that can be done in follow-on work, perhaps https://github.com/idaholab/moose/pull/16619
